app/src/main/java/com/humberto/tasky/core/domain/util/Result.kt

```diff
@@ -0,0 +1,19 @@
+package com.humberto.tasky.core.domain.util
+
+sealed interface Result<out D, out E: Error> {
+    data class Success<out D>(val data: D): Result<D, Nothing>
+    data class Error<out E: com.humberto.tasky.core.domain.util.Error>(val error: E): Result<Nothing, E>
+}
+
+inline fun <T, E: Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
+    return when(this) {
+        is Result.Error -> Result.Error(error)
+        is Result.Success -> Result.Success(map(data))
+    }
+}
+
+typealias EmptyResult<E> = Result<Unit, E>
+fun <T, E: Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
+    return map {  }
+}
```

Comment:
I've extended this error class a bit over time:
```
inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}
```
I like this more than unwrapping them in a when-expression, since you have one less indentation for the code that handles success and error cases:
```
repository.signUp(payload).onSuccess { responseData ->

}.onError { error -> 
    
}
```